app/queues/notifications/expiration-notification.server.ts

```diff
@@ -0,0 +1,152 @@
+// Third party modules
+import { Worker, Queue, UnrecoverableError } from 'bullmq';
+
+// Internal modules
+import { redis } from '~/lib/redis.server';
+import logger from '~/lib/logger.server';
+import { addNotification } from './notifications.server';
+import { prisma } from '~/db.server';
+
+// Types
+import type { NotificationData } from './notifications.server';
+
+export enum RecordType {
+  Certificate = 'certificate',
+  DnsRecord = 'record',
+}
+export interface ExpirationStatusPayload {
+  type: RecordType;
+}
+
+// name for the queue
+const expirationNotificationQueueName = 'expiration-notification';
+
+const updateNotificationStatus = (type: RecordType, id: number) => {
+  switch (type) {
+    case RecordType.Certificate:
+      return prisma.certificate.update({
+        where: {
+          id,
+        },
+        data: {
+          lastNotified: new Date(),
+        },
+      });
+    case RecordType.DnsRecord:
+      return prisma.record.update({
+        where: {
+          id,
+        },
+        data: {
+          lastNotified: new Date(),
+        },
+      });
+  }
+};
+
+// fetch records with by status
+const getRecordsByExpiration = (type: RecordType) => {
+  let monthFromToday = new Date();
+  monthFromToday.setMonth(monthFromToday.getMonth() + 1);
+
+  switch (type) {
+    case RecordType.Certificate:
+      return prisma.certificate.findMany({
+        where: {
+          validTo: {
+            lte: monthFromToday,
+          },
+          lastNotified: null,
+        },
+        select: {
+          user: true,
+          id: true,
+          lastNotified: true,
+        },
+      });
+    case RecordType.DnsRecord:
+      return prisma.record.findMany({
+        where: {
+          expiresAt: {
+            lte: monthFromToday,
+          },
+          lastNotified: null,
+        },
+        select: {
+          user: true,
+          id: true,
+          lastNotified: true,
+        },
+      });
+  }
+};
+// Queue initialization
+export const expirationNotificationQueue = new Queue<ExpirationStatusPayload>(
+  expirationNotificationQueueName,
+  {
+    connection: redis,
+  }
+);
+
+expirationNotificationQueue.on('error', (err) => {
+  logger.warn('Error running check for DNS Records/Certificates about to expire', err);
+});
+
+// function to add jobs
+export const addExpirationNotification = async (type: RecordType) => {
```

Comment:
> Actually, another question: do we even need to expose this to callers? Since we'll do this on a repeated cycle, should we call it ourselves internally on startup in this file, and no one ever needs to interact with it?

Currently, we are using the /dev route to initiate repeating jobs, so I need to expose it to call it in the front end. It should be automated on startup. 

I could look into ways to start the process automatically without user interaction.

Reply:
When you say "currently," it's being added in this PR.  I would skip putting it in `/dev` and do it on startup.  We have an `init()` model now (see how SAML and DNS do their startup logic once).